In [17]:
import pandas as pd
import statistics as stats
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [18]:
df= pd.read_csv("spambase.csv")

In [19]:
dataset = pd.read_csv

In [20]:
df.head()

Unnamed: 0  word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0           0            0.00               0.64           0.64           0.0   
1           1            0.21               0.28           0.50           0.0   
2           2            0.06               0.00           0.71           0.0   
3           3            0.00               0.00           0.00           0.0   
4           4            0.00               0.00           0.00           0.0   

  word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0          0.32            0.00              0.00                0.00   
1          0.14            0.28              0.21                0.07   
2          1.23            0.19              0.19                0.12   
3          0.63            0.00              0.31                0.63   
4          0.63            0.00              0.31                0.63   

   word_freq_order  ...  char_freq_;  char_freq_(  char_freq_[  char_freq_!  \
0             0.00  ...         0.00        0.000          0.0        0.778   
1             0.00  ...         0.00        0.132          0.0        0.372   
2             0.64  ...         0.01        0.143          0.0        0.276   
3             0.31  ...         0.00        0.137          0.0        0.137   
4             0.31  ...         0.00        0.135          0.0        0.135   

   char_freq_$  char_freq_#  capital_run_length_average  \
0        0.000        0.000                       3.756   
1        0.180        0.048                       5.114   
2        0.184        0.010                       9.821   
3        0.000        0.000                       3.537   
4        0.000        0.000                       3.537   

   capital_run_length_longest  capital_run_length_total  is_spam  
0                          61                       278     True  
1                         101                      1028     True  
2                         485                      2259     True  
3                          40                       191     True  
4                          40                       191     True  

[5 rows x 59 columns]

In [21]:
df.shape

(4601, 59)

 Dataset contains 4601 observations (rows) and 59 features (columns). The number of attributes is 58 (57 continuous and 1 nominal class label).
 Additional information obtained from "spambase_documentation" shows that spam constitutes 39.4% (1813 emails), while 2788 email are not spam (60.6%). 
 The attributes are continuous (attributes of type word_freq_WORD, attributes of type char_freq_CHAR and attributes of type capital_run_lenght) and also categorical (last feature -indicator "is_spam" which is a binary attribute)

This data shows the characteristics of emails based on their attributes Iword_freq,char_freq and capital_run_lenght). This characteristic is very important for identifying spam ( due to repeated charcters of words or frequent use of capital letters we can ditinguish spam from regular email).
We can also compare the diffrences between spam and non-spam. 

I would like to rename first column "Unnamed:0" to "ID"


In [22]:
df = df.rename(columns={"Unnamed: 0":"ID"})

In [23]:
df.head()

ID  word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0   0            0.00               0.64           0.64           0.0   
1   1            0.21               0.28           0.50           0.0   
2   2            0.06               0.00           0.71           0.0   
3   3            0.00               0.00           0.00           0.0   
4   4            0.00               0.00           0.00           0.0   

  word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0          0.32            0.00              0.00                0.00   
1          0.14            0.28              0.21                0.07   
2          1.23            0.19              0.19                0.12   
3          0.63            0.00              0.31                0.63   
4          0.63            0.00              0.31                0.63   

   word_freq_order  ...  char_freq_;  char_freq_(  char_freq_[  char_freq_!  \
0             0.00  ...         0.00        0.000          0.0        0.778   
1             0.00  ...         0.00        0.132          0.0        0.372   
2             0.64  ...         0.01        0.143          0.0        0.276   
3             0.31  ...         0.00        0.137          0.0        0.137   
4             0.31  ...         0.00        0.135          0.0        0.135   

   char_freq_$  char_freq_#  capital_run_length_average  \
0        0.000        0.000                       3.756   
1        0.180        0.048                       5.114   
2        0.184        0.010                       9.821   
3        0.000        0.000                       3.537   
4        0.000        0.000                       3.537   

   capital_run_length_longest  capital_run_length_total  is_spam  
0                          61                       278     True  
1                         101                      1028     True  
2                         485                      2259     True  
3                          40                       191     True  
4                          40                       191     True  

[5 rows x 59 columns]

I decided to remove column "ID" because this column is not relevant

In [24]:
df = df.drop(["ID"],axis=1)

In [25]:
df.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

  word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0          0.32            0.00              0.00                0.00   
1          0.14            0.28              0.21                0.07   
2          1.23            0.19              0.19                0.12   
3          0.63            0.00              0.31                0.63   
4          0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  is_spam  
0                       278     True  
1                      1028     True  
2                      2259     True  
3                       191     True  
4                       191     True  

[5 rows x 58 columns]

In [26]:
df.shape

(4601, 58)

Number of columns decreased from 59 to 58)

In [27]:
import seaborn as sns

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 58 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   word_freq_make              4601 non-null   float64
 1   word_freq_address           4601 non-null   float64
 2   word_freq_all               4595 non-null   float64
 3   word_freq_3d                4599 non-null   float64
 4   word_freq_our               4601 non-null   object 
 5   word_freq_over              4600 non-null   float64
 6   word_freq_remove            4601 non-null   float64
 7   word_freq_internet          4586 non-null   float64
 8   word_freq_order             4601 non-null   float64
 9   word_freq_mail              4601 non-null   float64
 10  word_freq_receive           4601 non-null   float64
 11  word_freq_will              4601 non-null   float64
 12  word_freq_people            4601 non-null   float64
 13  word_freq_report            4601 

In [30]:
df.isnull().sum()

word_freq_make                  0
word_freq_address               0
word_freq_all                   6
word_freq_3d                    2
word_freq_our                   0
word_freq_over                  1
word_freq_remove                0
word_freq_internet             15
word_freq_order                 0
word_freq_mail                  0
word_freq_receive               0
word_freq_will                  0
word_freq_people                0
word_freq_report                0
word_freq_addresses             9
word_freq_free                  0
word_freq_business              0
word_freq_email                 0
word_freq_you                   0
word_freq_credit                0
word_freq_your                  0
word_freq_font                  0
word_freq_000                   0
word_freq_money                 0
word_freq_hp                    0
word_freq_hpl                   0
word_freq_george                0
word_freq_650                   0
word_freq_lab                   0
word_freq_labs

Using this formula (df.isnull().sum()) I got information about missing data

In [32]:
df["word_freq_all"].unique()

array([0.64, 0.5 , 0.71, 0.  , 0.46, 0.77, 0.25, 0.34, 1.42, 0.42, 0.55,
       0.1 , 3.03, 2.56, 0.48, 1.66, 1.05, 0.18, 0.99, 0.26, 1.09, 0.73,
       1.26, 0.45, 1.68, 0.47, 0.12, 0.66, 0.61, 0.53, 0.89, 0.94, 1.77,
       0.37, 0.2 , 0.23, 0.72, 1.28, 0.38, 1.12, 0.29, 0.7 , 0.43, 1.24,
       0.76, 0.63, 1.02, 0.4 , 0.8 , 0.67, 0.41, 0.39, 0.33, 1.38, 0.27,
       0.56, 0.62, 0.68, 0.35, 2.38, 0.08, 1.  , 0.54, 1.32, 0.87, 1.08,
       1.07, 1.31, 0.22, 0.15, 0.52, 0.32, 0.16, 1.1 , 0.65, 1.43, 0.81,
       0.69, 0.58, 0.57, 0.14, 0.6 , 0.85, 1.63, 0.86, 0.17, 1.53, 1.29,
       0.19, 0.05, 1.45, 1.34, 1.35, 1.22, 0.31, 0.36, 1.33, 0.93, 0.59,
       0.11, 1.16, 0.03, 0.51, 0.49, 1.56, 1.11, 0.91, 1.44, 0.88, 1.92,
       0.09, 0.24, 1.25, 0.3 , 1.18, 0.28, 1.81, 0.84, 0.78, 0.75, 0.13,
       1.01, 0.9 , 1.47, 0.44, 0.98, 1.72, 1.23, 1.7 , 2.22, 1.65, 1.64,
       0.74, 0.95, 0.82, 0.97, 1.5 , 1.51, 1.17, 0.92, 1.13, 2.99, 0.04,
       1.62, 1.58, 1.9 , 1.83, 1.85, 1.61, 1.4 , 1.

Column word_freq_all : 6 missing data and no malformed values

In [33]:
df["word_freq_3d"].unique()

array([0.000e+00, 5.200e-01, 1.600e-01, 1.900e-01, 4.200e-01, 9.500e-01,
       1.700e-01, 5.030e+00, 7.180e+00, 1.363e+01, 8.100e-01, 1.160e+00,
       1.260e+00, 1.000e-01, 4.900e-01, 1.910e+00, 4.013e+01, 9.100e-01,
       9.160e+00, 4.310e+00, 4.281e+01, 1.916e+01, 6.000e-02, 3.546e+01,
       4.273e+01, 5.800e-01, 4.400e-01, 7.070e+00, 1.330e+00, 1.290e+00,
       2.100e-01, 1.973e+01, 5.700e-01, 6.000e-01, 4.000e-02, 1.350e+00,
       1.100e-01, 1.400e-01, 1.500e-01, 8.700e-01, 1.300e-01, 5.500e-01,
             nan, 3.100e-01])

Column word_freq_3d : 2 missing data (nan) and no malformed values

In [34]:
df["word_freq_over"].unique()

array([0.  , 0.28, 0.19, 0.32, 0.25, 0.35, 0.31, 0.05, 0.27, 0.92, 0.36,
       0.99, 0.12, 0.33, 0.73, 1.26,  nan, 0.46, 0.2 , 1.28, 0.76, 0.13,
       0.03, 0.6 , 0.64, 0.79, 0.16, 0.41, 0.4 , 0.66, 0.81, 0.39, 0.45,
       0.85, 0.29, 0.08, 0.78, 0.43, 0.87, 0.53, 0.56, 0.22, 0.42, 0.62,
       0.65, 0.54, 0.55, 1.12, 2.3 , 0.57, 0.1 , 0.17, 0.94, 0.7 , 0.18,
       0.34, 0.93, 0.51, 0.74, 0.44, 1.02, 0.48, 0.68, 0.11, 1.2 , 0.09,
       0.47, 0.61, 2.1 , 0.3 , 1.05, 1.29, 1.01, 0.84, 0.38, 0.98, 0.86,
       1.88, 1.32, 0.8 , 1.27, 1.1 , 0.23, 0.24, 0.15, 0.95, 0.63, 0.58,
       0.71, 0.37, 0.21, 0.67, 0.72, 1.14, 0.04, 0.5 , 0.14, 0.69, 0.91,
       0.9 , 0.59, 1.47, 1.61, 0.52, 1.25, 1.09, 1.86, 2.54, 0.77, 0.26,
       0.49, 1.49, 1.36, 1.4 , 1.21, 1.64, 1.04, 1.06, 1.11, 0.96, 2.43,
       0.06, 2.94, 0.75, 0.01, 0.07, 1.34, 0.02, 0.82, 1.19, 1.  , 1.07,
       2.63, 1.03, 5.88, 1.63, 3.44, 1.17, 0.89, 3.57, 1.57, 1.42])

Column word_freq_over : 1 missing data (nan) and no malformed values

In [35]:
df["word_freq_internet"].unique()

array([0.000e+00, 7.000e-02, 1.200e-01, 6.300e-01, 1.850e+00, 1.880e+00,
       3.500e-01,       nan, 2.000e-02, 2.450e+00, 5.200e-01, 1.810e+00,
       6.000e-02, 2.800e-01, 3.300e-01, 1.090e+00, 1.100e+00, 9.300e-01,
       3.600e-01, 1.260e+00, 2.500e-01, 1.450e+00, 1.860e+00, 2.600e-01,
       1.400e-01, 4.600e-01, 5.000e-01, 5.000e-02, 1.280e+00, 1.900e-01,
       8.400e-01, 1.200e+00, 4.000e-01, 2.980e+00, 1.310e+00, 2.900e-01,
       4.200e-01, 2.340e+00, 2.040e+00, 1.120e+00, 4.500e-01, 4.400e-01,
       1.700e-01, 1.300e-01, 3.200e-01, 5.900e-01, 4.300e-01, 2.240e+00,
       7.100e-01, 2.700e-01, 6.000e-01, 6.200e-01, 8.900e-01, 2.300e-01,
       1.050e+00, 6.400e-01, 1.800e-01, 1.410e+00, 3.400e-01, 3.700e-01,
       6.700e-01, 1.290e+00, 9.100e-01, 8.000e-02, 2.020e+00, 1.100e-01,
       5.800e-01, 1.360e+00, 2.100e-01, 1.070e+00, 4.800e-01, 2.530e+00,
       1.080e+00, 4.700e-01, 5.300e-01, 4.900e-01, 3.330e+00, 2.960e+00,
       2.820e+00, 4.620e+00, 2.000e-01, 1.000e-01, 

Column word_freq_internet : 15 missing data (nan) and no malformed values

In [36]:
df["word_freq_labs"].unique()

array([nan, '0.39', '0', '0.06', '0.66', '2.24', '0.24', '3.38', '0.97',
       '0.86', '0.62', '0.17', '0.11', '0.12', '0.1', '0.18', '2.04',
       '0.73', '0.58', '1.44', '0.26', '2.7', '3.57', '0.3', '2.77',
       '0.46', '1', '0.14', '2.17', '0.31', '0.32', '0.08', '1.06', '2',
       '0.89', '1.73', '0.42', '1.96', '0.48', '1.66', '0.52', '4.54',
       '0.54', '0.27', '0.9', '1.52', '1.02', '0.15', '0.55', '0.28',
       '1.33', '0.35', '1.11', '1.31', '0.44', '0.34', '0.93', '0.68',
       '0.25', '0.16', '0.07', '1.85', '0.82', '0.05', '1.81', '0.74',
       '0.91', '0.57', '0.51', '0.72', '0.5', '0.99', '0.96', '2.32',
       '0.21', '1.19', '1.08', '1.26', '1.38', '0.22', '0.09', '4.16',
       '1.17', '0.33', '0.63', '1.03', '1.23', '1.84', '0.2', '2.27',
       '0.87', '3.22', '0.61', '0.19', '0.45', '2.29', '0.53', '0.37',
       '1.28', '0.84', '0.75', '1.29', '0.38', '0.77', '1.58', '0.23',
       '4.76', '2.45', '0.7', '0.56', '2.19', '0.78', '0.92', '1.42',
       '1

Column word_freq_labs : 250 mising data (nan) and 1 malformed value (no)

In [37]:
df["word_freq_addresses"].unique()

array([0.  , 0.14, 1.75, 0.12, 0.18, 1.27,  nan, 0.3 , 0.26, 0.03, 2.31,
       0.05, 0.16, 0.34, 0.15, 0.17, 0.51, 0.45, 1.11, 0.62, 0.46, 0.07,
       0.19, 2.22, 1.69, 2.38, 2.09, 0.35, 2.21, 0.76, 0.39, 1.92, 1.93,
       0.04, 2.12, 0.06, 1.88, 0.66, 0.24, 0.13, 0.79, 2.16, 1.61, 0.09,
       0.31, 0.02, 0.25, 0.63, 1.46, 0.32, 1.42, 2.06, 1.  , 0.89, 1.47,
       0.75, 0.5 , 0.69, 0.2 , 1.57, 0.23, 0.47, 2.11, 2.05, 1.17, 1.5 ,
       1.51, 1.66, 0.37, 0.65, 1.07, 0.22, 1.1 , 1.81, 1.15, 1.14, 0.42,
       0.4 , 0.97, 1.04, 0.54, 0.53, 0.6 , 1.6 , 0.29, 0.55, 0.78, 0.77,
       1.02, 0.1 , 0.61, 0.28, 1.58, 0.86, 0.84, 0.41, 0.58, 0.59, 0.27,
       1.94, 4.41, 0.33, 0.68, 0.83, 0.71, 0.81, 2.86, 1.44, 1.7 , 0.08,
       1.36, 1.28, 0.01, 1.4 , 1.49, 1.16, 1.19, 2.24, 0.44])

Columns word_freq_addresses: 9 missing data and no malformed values

In [38]:
df["word_freq_direct"].unique()

array([0.  , 0.06, 0.96, 0.42, 1.59, 0.12, 0.48, 0.46, 0.16, 0.19, 0.14,
       0.15, 0.66, 0.91, 0.65, 0.44, 0.02, 0.08, 0.92, 1.07, 0.17, 0.1 ,
       0.36, 0.59, 0.11, 0.29, 0.68, 0.5 , 0.35, 0.53, 0.47, 0.13, 0.93,
       0.09, 0.97, 0.28, 0.03, 2.22, 0.63, 0.2 , 0.07, 0.34, 0.89, 0.64,
       0.31, 0.52, 0.51, 0.04, 0.74, 0.37, 1.19, 0.24, 0.49, 0.4 , 0.26,
       1.62, 0.27, 1.36, 0.32, 0.38, 0.33, 0.25, 1.  , 0.58, 0.61, 0.43,
       0.45, 0.55, 0.22, 2.04, 0.05, 2.11, 4.54, 0.76, 0.23, 0.39, 1.11,
       3.57, 1.31, 0.54, 1.85, 0.6 , 0.72, 0.73, 0.99, 0.21, 1.08, 1.38,
       4.16, 2.27, 0.87, 0.82, 1.28, 0.75, 1.58, 4.76, 2.77, 1.42, 1.01,
       4.34, 0.86, 0.62, 3.12, 1.56, 2.63, 1.72, 0.8 , 0.01, 0.56, 0.78,
       3.03, 2.12, 2.56, 1.17, 3.84, 1.2 , 2.  , 4.  , 0.9 , 0.85, 2.32,
       0.88, 1.44, 0.77, 0.67,  nan])

Columns word_freq_direct: 3 missing data and no malformed values

In [39]:
df["word_freq_cs"].unique()

array([0.  , 0.1 , 2.32, 0.54, 0.39, 7.14, 3.22, 0.51, 1.16, 0.13, 0.89,
       0.25, 0.76, 0.34, 1.38, 0.33, 0.08, 3.25, 0.59, 1.24, 0.18, 0.17,
       1.3 , 3.26, 0.31, 0.19, 0.65, 1.56, 3.12, 1.44, 1.62, 0.68, 4.75,
       1.29, 0.06, 2.31, 0.36, 0.04, 1.01, 0.26, 0.16, 2.  , 0.94, 1.4 ,
       0.29, 0.28, 0.71, 3.07, 0.03, 2.11, 0.2 , 0.61, 1.58, 0.35, 0.37,
       2.5 , 2.56, 0.56, 5.  , 0.46, 2.8 , 5.88, 2.22, 2.26, 1.7 , 1.35,
       1.92, 1.94, 1.72, 0.92, 0.8 ,  nan, 0.52, 1.26, 0.38, 0.98, 0.75,
       1.98, 0.32, 0.69, 0.07, 0.02, 0.93, 2.83, 1.69, 1.33, 0.84, 2.81,
       2.15, 0.86, 1.06, 0.81, 1.8 , 1.66, 1.28, 0.23, 0.62, 0.6 , 1.2 ,
       1.17, 0.44, 4.28, 2.69, 0.53, 1.88, 0.21, 1.02, 0.42, 4.76])

Column word_freq_cs: 14 missing data and no malformed values

In [40]:
df["word_freq_meeting"].unique()

array([0.000e+00, 1.300e-01, 6.000e-02, 2.300e-01, 3.400e-01, 1.100e-01,
       1.800e-01, 3.000e-02, 2.500e-01,       nan, 4.500e-01, 3.200e-01,
       3.500e-01, 3.800e-01, 3.700e-01, 3.600e-01, 4.080e+00, 7.300e-01,
       9.000e-01, 3.700e+00, 1.850e+00, 2.890e+00, 1.600e+00, 1.200e-01,
       1.880e+00, 8.000e-02, 2.120e+00, 9.100e-01, 6.700e-01, 2.000e-02,
       1.900e+00, 9.600e-01, 7.000e-02, 1.750e+00, 1.810e+00, 9.090e+00,
       1.000e-02, 9.200e-01, 1.340e+00, 3.220e+00, 5.120e+00, 2.230e+00,
       6.000e-01, 9.700e-01, 1.620e+00, 5.400e-01, 2.430e+00, 1.020e+00,
       7.600e-01, 9.300e-01, 3.900e-01, 8.900e-01, 1.100e+00, 2.800e-01,
       4.000e-01, 4.540e+00, 5.000e+00, 3.840e+00, 1.050e+00, 5.300e-01,
       6.600e-01, 8.500e-01, 1.860e+00, 2.690e+00, 1.380e+00, 7.100e-01,
       1.040e+00, 7.140e+00, 2.220e+00, 1.440e+00, 2.020e+00, 5.800e-01,
       7.800e-01, 3.440e+00, 1.090e+00, 2.400e+00, 2.200e-01, 8.700e-01,
       6.060e+00, 4.800e-01, 4.000e-02, 5.700e-01, 

Column word_freq_meeting: 1 missing data and no malformed values

In [41]:
df["word_freq_table"].unique()

array([0.  , 0.04, 0.34, 0.46, 0.02, 0.03,  nan, 0.06, 0.09, 0.11, 0.37,
       0.16, 0.27, 1.02, 1.91, 0.65, 0.19, 0.18, 0.74, 0.86, 0.05, 0.28,
       0.25, 0.39, 0.01, 1.6 , 2.17, 0.12, 0.14, 0.93, 1.51, 0.88, 2.12,
       0.73, 0.61, 0.81, 0.72, 0.51, 0.8 ])

Column word_freq_table: 40 missing data and no malformed values